Add benchmarks

chrisblossom opened the pull request:
This is not necessarily meant to be merged as is, but meant to try and answer https://github.com/sindresorhus/del/issues/60.

sindresorhus commented on benchmark.js:
```diff
@@ -0,0 +1,70 @@
+
+const concurrency = [1, 3, 5, 10, 15, 20, 50, 100, Infinity];
+
+concurrency.forEach(num => {
```
I prefer `for-of` loops whenever possible. Also don't use abbreviations `num` => `number`.

Actually, this should be:
```suggestion
concurrencies.forEach(concurrency => {
```

chrisblossom replied:
Updated

sindresorhus commented on benchmark.js:
```diff
@@ -0,0 +1,70 @@
+			// https://github.com/bestiejs/benchmark.js/issues/136
+			createFixtures();
+
+			// Async await was giving too many errors. stick with standard promises
```
Huh? What does this mean? Async/await should work any differently.

chrisblossom replied:
```js
suite.add({
	name,
	defer: true,
    // if you remove setup() the benchmark works as expected
	setup() {
	},
	async fn(deferred) {
		// Can't use setup because it isn't called after every defer
		// https://github.com/bestiejs/benchmark.js/issues/136
		createFixtures();

		// Async await was giving too many errors. stick with standard promises
		await del(['**/*'], {
			cwd: tempDir,
			concurrency,
			// eslint-disable-next-line promise/prefer-await-to-then
		}).then(removedFiles => {
			if (removedFiles.length !== fixtures.length) {
				const error = new Error(
					`"${name}": files removed: ${removedFiles.length}, expected: ${fixtures.length}`,
				);

				console.error(error);

				del.sync(tempDir, { cwd: tempDir, force: true });

				// eslint-disable-next-line unicorn/no-process-exit
				process.exit(1);
			}

			deferred.resolve();
		});
	},
});
```

Gives the error:
```js
del on  add-benchmarks [⇡!] is 📦 v5.0.0 via ⬢ v10.16.0 took 6s
➜ node benchmark.js
undefined:9
			await del(['**/*'], {
			^^^^^

SyntaxError: await is only valid in async function
    at Function (<anonymous>)
    at createCompiled (/Users/chris/github/del/node_modules/benchmark/benchmark.js:1725:16)
    at clock (/Users/chris/github/del/node_modules/benchmark/benchmark.js:1608:58)
    at clock (/Users/chris/github/del/node_modules/benchmark/benchmark.js:1818:20)
    at new Deferred (/Users/chris/github/del/node_modules/benchmark/benchmark.js:405:7)
    at Deferred (/Users/chris/github/del/node_modules/benchmark/benchmark.js:402:16)
    at Benchmark.run (/Users/chris/github/del/node_modules/benchmark/benchmark.js:2112:13)
    at execute (/Users/chris/github/del/node_modules/benchmark/benchmark.js:860:74)
    at Timeout._onTimeout (/Users/chris/github/del/node_modules/lodash/lodash.js:2750:43)
    at ontimeout (timers.js:436:11)
```

sindresorhus replied:
Weird. I already commented on this, but it's gone now... Anyway, I think you're just missing the `async` keyword for the method. Then it worked for me at least.

sindresorhus commented:
Results for me on Node.js 12.6.0 and macOS 10.14:

```
concurrency: 1 x 2.09 ops/sec ±11.43% (15 runs sampled)
concurrency: 3 x 2.86 ops/sec ±8.24% (18 runs sampled)
concurrency: 5 x 3.18 ops/sec ±14.31% (19 runs sampled)
concurrency: 10 x 3.60 ops/sec ±9.86% (22 runs sampled)
concurrency: 15 x 3.77 ops/sec ±7.95% (23 runs sampled)
concurrency: 20 x 4.05 ops/sec ±9.13% (25 runs sampled)
concurrency: 50 x 4.03 ops/sec ±9.58% (25 runs sampled)
concurrency: 100 x 3.67 ops/sec ±10.94% (23 runs sampled)
concurrency: Infinity x 3.84 ops/sec ±6.76% (23 runs sampled)
Fastest is concurrency: 20,concurrency: 50,concurrency: 15,concurrency: 100
```

How about you?

chrisblossom commented:
On my desktop:
```
del on  add-benchmarks is 📦 v5.0.0 via ⬢ v12.6.0 took 16s
➜ npx envinfo --system --binaries
npx: installed 1 in 0.753s

  System:
    OS: macOS 10.14.5
    CPU: (16) x64 Intel(R) Xeon(R) W-2140B CPU @ 3.20GHz
    Memory: 2.31 GB / 32.00 GB
    Shell: 5.7.1 - /usr/local/bin/zsh
  Binaries:
    Node: 12.6.0 - ~/.nvm/versions/node/v12.6.0/bin/node
    Yarn: 1.17.0 - /usr/local/bin/yarn
    npm: 6.9.0 - ~/.nvm/versions/node/v12.6.0/bin/npm


del on  add-benchmarks is 📦 v5.0.0 via ⬢ v12.6.0
➜ node benchmark.js
concurrency: 1 x 3.07 ops/sec ±2.17% (20 runs sampled)
concurrency: 3 x 5.14 ops/sec ±1.15% (29 runs sampled)
concurrency: 5 x 6.10 ops/sec ±0.55% (34 runs sampled)
concurrency: 10 x 5.67 ops/sec ±2.16% (31 runs sampled)
concurrency: 15 x 6.42 ops/sec ±0.78% (35 runs sampled)
concurrency: 20 x 6.45 ops/sec ±0.42% (35 runs sampled)
concurrency: 50 x 6.41 ops/sec ±0.96% (35 runs sampled)
concurrency: 100 x 6.76 ops/sec ±0.70% (37 runs sampled)
concurrency: Infinity x 6.08 ops/sec ±0.51% (34 runs sampled)
Fastest is concurrency: 100
```

chrisblossom commented:
Also, if you have any ideas how to remove `createFixtures()` from the benchmark I'm all ears. Only way I can think of to do this is to remove `benchmark.js` and manually calculate each iteration.

sindresorhus commented on benchmark.js:
```diff
@@ -0,0 +1,69 @@
+
+suite
+	.on('cycle', event => {
+		console.log(String(event.target));
```
console.log already stringifies.

```suggestion
		console.log(event.target);
```

chrisblossom replied:
I copied and pasted that code from an example `benchmark.js` had on their website. I just tried changing it and the console is spammed with:

```js
Benchmark {
  options: {
    async: false,
    defer: true,
    delay: 0.005,
    id: undefined,
    initCount: 1,
    maxTime: 5,
    minSamples: 5,
    minTime: 0.05,
    name: 'concurrency: 1',
    onAbort: undefined,
    onComplete: undefined,
    onCycle: undefined,
    onError: undefined,
    onReset: undefined,
    onStart: undefined,
    fn: [Function: fn]
  },
  async: false,
  defer: true,
  delay: 0.005,
  initCount: 1,
  maxTime: 5,
  minSamples: 5,
  minTime: 0.05,
  name: 'concurrency: 1',
  fn: [Function: fn],
  id: 1,
  stats: {
    moe: 0.01129889530845934,
    rme: 3.3082535827432205,
    sem: 0.005377865449052517,
    deviation: 0.02344157202437891,
    mean: 0.34153655473684214,
    sample: [
      0.397719688, 0.388559305,
      0.379714549, 0.343945989,
      0.341582504, 0.333736164,
       0.35213655, 0.321157255,
      0.327945198,  0.34182276,
      0.320502646, 0.319912421,
      0.313889747, 0.332924082,
      0.334101052,   0.3319376,
       0.33097862, 0.349683892,
      0.326944518
    ],
    variance: 0.0005495072989741439
  },
  times: {
    cycle: 0.34153655473684214,
    elapsed: 6.796,
    period: 0.34153655473684214,
    timeStamp: 1566434288783
  },
  _timerId: Timeout {
    _idleTimeout: 5,
    _idlePrev: null,
    _idleNext: null,
    _idleStart: 96,
    _onTimeout: [Function],
    _timerArgs: undefined,
    _repeat: null,
    _destroyed: false,
    [Symbol(refed)]: null,
    [Symbol(asyncId)]: 2,
    [Symbol(triggerId)]: 1
  },
  events: { complete: [] },
  running: false,
  count: 1,
  compiled: [Function: anonymous],
  cycles: 1,
  hz: 2.927944274575562
}
```

sindresorhus commented on benchmark.js:
```diff
@@ -0,0 +1,69 @@
+		defer: true,
+		fn(deferred) {
+			// Can't use setup because it isn't called after every defer
+			// https://github.com/bestiejs/benchmark.js/issues/136
```
The referenced issue is closed now.

chrisblossom replied:
It is just a reference to why setup can't be used. Want me to remove it?

sindresorhus commented:
Seems like 100 would be a good default. Maybe you could add 200, 300, 400, 500, 1000 too? Just to check.

chrisblossom commented:
From my laptop:
```
concurrency: 1 x 3.04 ops/sec ±2.91% (19 runs sampled)
concurrency: 3 x 4.84 ops/sec ±0.86% (28 runs sampled)
concurrency: 5 x 5.35 ops/sec ±1.81% (30 runs sampled)
concurrency: 10 x 5.16 ops/sec ±2.28% (29 runs sampled)
concurrency: 15 x 5.30 ops/sec ±2.94% (30 runs sampled)
concurrency: 20 x 5.57 ops/sec ±1.64% (31 runs sampled)
concurrency: 50 x 5.79 ops/sec ±0.86% (32 runs sampled)
concurrency: 100 x 5.87 ops/sec ±1.46% (32 runs sampled)
concurrency: 200 x 6.06 ops/sec ±0.93% (33 runs sampled)
concurrency: 300 x 6.10 ops/sec ±1.94% (34 runs sampled)
concurrency: 400 x 6.11 ops/sec ±0.83% (34 runs sampled)
concurrency: 500 x 6.01 ops/sec ±0.91% (33 runs sampled)
concurrency: 1000 x 5.72 ops/sec ±3.55% (33 runs sampled)
concurrency: Infinity x 5.49 ops/sec ±0.59% (31 runs sampled)
Fastest is concurrency: 400,concurrency: 200,concurrency: 300
```

sindresorhus commented on benchmark.js:
```diff
@@ -0,0 +1,69 @@
+				cwd: tempDir,
+				concurrency
+				// eslint-disable-next-line promise/prefer-await-to-then
+			}).then(removedFiles => {
```
Use `await` instead of `.then()`.

chrisblossom commented:
I've resolved the merge conflict and updated to async await. I added the line that demonstrates why I was using normal promises. Let me know if you want me to revert the async await change or remove the `setup() {}` line.

sindresorhus commented:
I just forgot to remove the setup method before merging as I had intended to: https://github.com/sindresorhus/del/commit/728cb7cbea6bcb94b1481c96ec40a7465cdcf3e6
